kinesis/src/main/scala/akka/stream/alpakka/kinesis/KinesisFlowStage.scala

```diff
@@ -0,0 +1,184 @@
+}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+import scala.language.postfixOps
+
+private[kinesis] final class KinesisFlowStage(
+    streamName: String,
+    maxRetries: Int,
+    backoffStrategy: RetryBackoffStrategy,
+    retryInitialTimeout: FiniteDuration
+)(implicit kinesisClient: AmazonKinesisAsync)
+    extends GraphStage[FlowShape[Seq[PutRecordsRequestEntry], Future[PutRecordsResult]]] {
+
+  private val in = Inlet[Seq[PutRecordsRequestEntry]]("records")
+  private val out = Outlet[Future[PutRecordsResult]]("result")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with StageLogging {
+
+      private val retryBaseInMillis = retryInitialTimeout.toMillis
+
+      private var completionState: Option[Try[Unit]] = None
+
+      private val pendingRequests: mutable.Queue[(Int, Seq[PutRecordsRequestEntry])] = mutable.Queue.empty
+      private var resultCallback: AsyncCallback[Result] = _
+      private var inFlight = 0
+
+      private var retryToken = 0
+      private val waitingRetries: mutable.HashMap[Int, Result] = mutable.HashMap.empty
+
+      private def tryToExecute() =
+        if (pendingRequests.nonEmpty && isAvailable(out)) {
+          log.debug("Executing PutRecords call")
+          inFlight += 1
+          val (attempt, records) = pendingRequests.dequeue()
+          push(
+            out,
+            putRecords(
+              streamName,
+              records,
+              recordsToRetry => resultCallback.invoke(Result(attempt, recordsToRetry))
+            )
+          )
+        }
+
+      private def handleResult(result: Result): Unit = result match {
+        case Result(_, Nil) =>
+          log.debug("PutRecords call finished successfully")
+          inFlight -= 1
+          tryToExecute()
+          if (!hasBeenPulled(in)) tryPull(in)
+          checkForCompletion()
+        case Result(attempt, errors) if attempt > maxRetries =>
+          log.debug("PutRecords call finished with partial errors after {} attempts", attempt)
+          failStage(ErrorPublishingRecords(attempt, errors.map(_._1)))
+        case x @ Result(attempt, _) =>
+          log.debug("PutRecords call finished with partial errors; scheduling retry")
+          inFlight -= 1
+          waitingRetries.put(retryToken, x)
+          scheduleOnce(retryToken, backoffStrategy match {
+            case Exponential => scala.math.pow(retryBaseInMillis, attempt).toInt millis
+            case Lineal => retryInitialTimeout * attempt
+          })
+          retryToken += 1
+      }
+
+      private def checkForCompletion() =
+        if (inFlight <= 0 && pendingRequests.isEmpty && waitingRetries.isEmpty && isClosed(in)) {
```

Comment:
should `inFlight` ever be able to become `< 0`? Perhaps log a warning then?

Reply:
No, it cannot happen. I'll change it to `== 0`.